### PR TITLE
Implement Prepare Mechanic

### PR DESCRIPTION
**Pull request by nandmp**
#14329

### CODE REVIEW

**Comment by wakame1367** [comment]
FYI: I closed #15599 in favor of this PR — your exile-based model matches CR 722.3c, mine didn't.

If useful, I have a small gameplay test suite for prepare I'm happy to adapt against your branch: cast-from-exile timing (instant vs sorcery faces), unprepare-on-cast, re-prepare behavior, plus a corner case Grath suggested in #15599 (a prepared permanent becoming a copy of a different prepare card must not change the spell copy in exile). Just say the word.

**Comment by nandmp** [comment]
Sure, go ahead.
Someone (if not you) also suggested some more edge cases here: https://www.reddit.com/r/EDH/comments/1sz7b43/prepared_mechanic_deep_dive_a_comprehensive_guide/

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+        addCard(Zone.BATTLEFIELD, playerA, "Mountain", 2);
+
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, CREATURE);
+        waitStackResolved(1, PhaseStep.PRECOMBAT_MAIN);
+        runCode("try to prepare again", 1, PhaseStep.PRECOMBAT_MAIN, playerA, (info, player, game) -> {
+            Permanent permanent = game.getBattlefield().getAllActivePermanents(playerA.getId()).stream()
+                    .filter(p -> p.getName().equals(CREATURE)).findFirst().orElse(null);
+            Assert.assertNotNull(permanent);
+            Assert.assertTrue(permanent.isPrepared());
+            permanent.setPrepared(true, game);
```
Comment (matoro) [inline Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java]: I think this test should be done using a legal card rather than running code that introspects the game state

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        addCard(Zone.HAND, playerA, CREATURE);
+        addCard(Zone.BATTLEFIELD, playerA, "Mountain", 2);
+
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, CREATURE);
+        waitStackResolved(1, PhaseStep.PRECOMBAT_MAIN);
+        runCode("make unprepared", 1, PhaseStep.PRECOMBAT_MAIN, playerA, (info, player, game) -> {
+            Permanent permanent = game.getBattlefield().getAllActivePermanents(playerA.getId()).stream()
+                    .filter(p -> p.getName().equals(CREATURE)).findFirst().orElse(null);
+            Assert.assertNotNull(permanent);
+            permanent.setPrepared(false, game);
```
Comment (matoro) [inline Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java]: same.  you can reuse Tomekeeper for this

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+                "Exile a creature card from your graveyard");
+        setChoice(playerA, "Grizzly Bears");
+
+        setStrictChooseMode(true);
+        setStopAt(1, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertGraveyardCount(playerA, "Grizzly Bears", 0);
+        assertExileCount(playerA, "Grizzly Bears", 2);
+        assertExileCount(playerA, "Lluwen, Exchange Student", 1);
```
Comment (matoro) [inline Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java]: shouldn't this be the prepare spell ([[Pest Friend]]) in exile?  why is this passing?

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        waitStackResolved(3, PhaseStep.PRECOMBAT_MAIN);
+        activateAbility(3, PhaseStep.PRECOMBAT_MAIN, playerA,
+                "{3}, {T}: Target creature becomes prepared", CREATURE);
+
+        setStrictChooseMode(true);
+        setStopAt(3, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertPermanentCount(playerA, "Treasure Token", 1);
+        assertExileCount(playerA, CREATURE, 1);
```
Comment (matoro) [inline Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java]: same

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "Nanogene Conversion", "Emeritus of Woe");
+
+        setStrictChooseMode(true);
+        setStopAt(1, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertPermanentCount(playerA, "Emeritus of Woe", 2);
+        // Only the original Emeritus supplied a prepared copy. The Llanowar
+        // Elves did not become prepared retroactively after copying it.
+        assertExileCount(playerA, "Emeritus of Woe", 1);
```
Comment (matoro) [inline Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/PrepareTest.java]: same

**Review by matoro — COMMENTED** [review]
other than these comments this looks fully accurate to me

**Review by matoro — APPROVED** [review]
I don't have merge authority, but this looks as good as it's gonna get to me.

**Comment by JayDi85** [comment]
It contains dozen hacks and non-standard code - so need additional reviews. I'll look at it later.

**Comment by jeffwadsworth** [comment]
There should be a test for canceling the casting of the exiled copy and confirming the exiled copy is still available and the permanent remains prepared.

**Comment by Ninja058** [comment]
Any updates regarding this?

**Review by Grath — APPROVED** [review]
This looks like the most reasonable approach to implementing Prepare; unifies handling for Adventure, Omen, and Prepare, handles the update to copy cleanup SBAs, and the changes are thoroughly tested. The hacky weirdness has already been cleaned up. I think this can be safely merged at this point.

**Comment by Grath** [comment]
The dozen hacks and non-standard code have been fixed up since the [don't merge], removing that and merging this.
